Simplify Hypershift Logging Operator proposal. (#1625)
openshift/enhancements: enhancements/cluster-logging/hypershift-audit-logs.md

## Summary

In a self contained cluster, users can forward application, infrastructure and audit logs.
In a hypershift hosted cluster, API server audit logs are not exposed in the hosted cluster or on the data plane.
They are generated in Hosted Control Plane (HCP) pods on a management cluster and forwarded via HTTP web-hook.

This proposal deploys the existing Cluster Logging Operator (CLO) on management and hosted clusters.
The CLO provides a ClusterLogForwarder resource (CLF) that collects and forwards logs.
On the hosted cluster, a CLF can be configured to receive audit logs from the management cluster
and forward them to customer-configured destinations.

On the management cluster, this proposal adds a "Hypershift Logging Operator" (HLO).

> The HLO watches _data-plane_ CLF resources, when it sees one configured to receive audit logs it updates a _control-plane_ CLF to do the forwarding.

Reviewer: IMO this looks like the need to make use of annotations and labels.  @alanconway this looks like a case where either an admission hook or serverside apply could update the CLF with an http receiver requiring the customer to only say "I need audit logs".  HTTP receivers currently only accept audit logs though the HLO could use some convetion based upon the annotation or label applied by the customer to know "which" is for audit events

> The HLO watches _data-plane_ CLF resources, when it sees one configured to receive audit logs it updates a _control-plane_ CLF to do the forwarding.

Author: Customer doesn't have to configure anything on the HLO. I was suggesting that customer configures their CLF with a HTTP receiver that has a "magic" name, and that can be the trigger to configure the HLO. Could alternatively use annotations, that might be more explicit, e.g. an annotation on the CLF with the output name as a value. I can make the change if you agree @jewzaam @jcantrill

> The HLO watches _data-plane_ CLF resources, when it sees one configured to receive audit logs it updates a _control-plane_ CLF to do the forwarding.

Author: I'm rewriting with the use of an annotation -  I like that its more explict/less magical. Available for review shortly...

> The control plane CLF is also used to forward to Splunk and optionally CloudWatch from the control plane directly.

Author: This doc is aimed at the ROSA HCP case but the forwarding feature is intended to be more general. I'll add a note to clarify that and give an example of control-plane only forwarding.

#### As a service provider, I want to forward selected API audit logs to my chosen destinations

- configure the SD internal audit policy
- configure API audit policy and forwarding using name-spaced logging customer resources in the management cluster.
  - for example: Red Hat Service Delivery team forwards audit logs to internal Splunk destination.
- independent of user configuration.

### Goals

- Multiple audit policies can be defined in a management cluster, independent of hosted cluster policies.
- Multiple audit policies can be defined in a hosted cluster, independent of the management cluster policies.
  See [api-audit-log-policy](/api-audit-log-policy.md) for how the CLO will support this.
- Data plane (customer) requests audit log forwarding by modifying data-plane CLF configuration, forwarding from control plane starts automatically.

### Non-Goals

- This proposal only covers API audit logs.
  Support for application, infrastructure or node audit logs could be added but is not discussed here.

## Proposal

### Workflow Description

#### Hosted Cluster, data plane

1. Install the Cluster Logging Operator (CLO).
1. Create a `ClusterLogForwarder` with a HTTP receiver.

>    - Receiver must be identified with a special input name e.g. `hypershiftAPIAudit`

Author: I'll make  it an official statement.

## Summary

In a self contained cluster, users can forward application, infrastructure and audit logs.
In a hypershift hosted cluster, API server audit logs are not exposed in the hosted cluster or on the data plane.
They are generated in Hosted Control Plane (HCP) pods on a management cluster and forwarded via HTTP web-hook.

This proposal deploys the existing Cluster Logging Operator (CLO) on management and hosted clusters.
The CLO provides a ClusterLogForwarder resource (CLF) that collects and forwards logs.
On the hosted cluster, a CLF can be configured to receive audit logs from the management cluster
and forward them to customer-configured destinations.

On the management cluster, this proposal adds a "Hypershift Logging Operator" (HLO).

> The HLO watches _data-plane_ CLF resources, when it sees one configured to receive audit logs it updates a _control_plane_ CLF to do the forwarding.

Reviewer 2: We were using the HLF custom resource for HLO in the initial implementation. Does this mean we are going to deprecate that and use HLO to watch the CLF directly?

Reviewer 2: ok, seems this answered below.

> The control plane CLF is also used to forward to Splunk and optionally CloudWatch from the control plane directly.

Reviewer 3: This doesn't have to be true, is this enhancement written to be generic or specific to managed ocp?

```suggestion
The control plane CLF can be configured to forward to additional supported outputs from the control plane directly.
```

Author: This enhancement is written for managed OCP. The CLO features could be used in other settings, but the HLO is intended specifically for the HCP situation. We could generalize it more but I don't know of any desire to release it outside of OCP at the moment.

Reviewer 3: HCP isn't managed only, there are self-managed HCP deployments.  I don't expect what is built here to be only for managed.  It would be in the regular CLO release payload right?  Documented?

Author: +1 I'll be more explicit about what is just a "suggestion" or "example" of using existing CLF features for the ROSA HCP case, vs. what's being proposed as new functionality.

> The control plane CLF can be configured to forward to additional supported outputs from the control plane directly.

Reviewer 2: Also there was a CR CLFT on management cluster, this is going to be replaced by CLF as well?

Reviewer 2: ok, seems this answered below.

Author: +1. Goal here is to reduce the number of new entities and new functionality.

  - for example: Red Hat Service Delivery team forwards audit logs to internal Splunk destination.
- independent of user configuration.

### Goals

- Multiple audit policies can be defined in a management cluster, independent of hosted cluster policies.
- Multiple audit policies can be defined in a hosted cluster, independent of the management cluster policies.
  See [api-audit-log-policy](/api-audit-log-policy.md) for how the CLO will support this.
- Allow some forwarding in the _management_ cluster to be configured from the _hosted_ cluster.
- Keep direct control of management cluster forwarding on the management cluster.
  - hosted cluster has no direct access to control plane configuration.
  - data plane resources can request forwarding by the management cluster.
  - management cluster HLO automatically validates data-plane requests, and enables them if they are safe.
- Data plane (customer) requests audit log forwarding by modifying data-plane CLF configuration, forwarding from control plane starts automatically.

### Non-Goals

- This proposal only covers API audit logs. No change to application, infrastructure or node audit logs.

> - Management cluster is not _required_ to honour all possible forwarding configurations allowed by CLF. \

Reviewer 3: Will CLF on MC support everything?

Author: Yep. I wrote the original constraint because we were using the MC CLF to support both guest and host personas, which was weird, so I wanted to anticipate we might have limitations. Now the guest and host CLF are independent (except for a HTTP forwarding link which isn't a strong coupling) so there's no reason to expect special limitations on either side.

> 1. Create a `ClusterLogForwarder` with a HTTP intupt.

Reviewer 3: ```suggestion
1. Create a `ClusterLogForwarder` with a HTTP input.
```

> 1. Create a `ClusterLogForwarder` with a HTTP intupt.

Reviewer 4: Are these installed / setup for every hostedcluster , or opt in by a customer at install time or day-2 ?

Author: The data-plane CLF is opt-in by the customer on day 2. The control-plane CLF is created automatically by the HLO in response to the customer CLF existing and having the relevant configuration.

> 1. Create a `ClusterLogForwarder` for each HCP to forward to desired outputs.

Reviewer 2: I remember at the beginning, we have limitation that there can be only one CLF to be consumed by CLO.
So with the latest design, there will be no such limitation, and each CLF can define the separated forwording rule and work together?

Author: +1 @bmeng this helps us avoid the CLFT or similar on the management side. Instead of a more complex template that includes splunk + http forwarding, we can separate distinct forwarding tasks into separate CLF configurations. We don't _have_ to separate them, but it makes the individual configurations simpler and provides better isolation between the tasks in case a forwarder fails.

> 1. Create a `ClusterLogForwarder` for each HCP to forward to Splunk and management-side CloudWatch.

Reviewer 3: ```suggestion
1. Create a `ClusterLogForwarder` for each HCP to forward to desired outputs.
```

>    - When the HLO detects a _control plane_ CLF requesting audit log forwarding, it updates
>      the _data plane_ CLF for that HCP to forward audit logs via a HTTP output to the _control plane_ CLF.

Reviewer 3: ```suggestion
   - When the HLO detects a _data plane_ CLF requesting audit log forwarding, it updates
     the _control plane_ CLF for that HCP to forward audit logs via a HTTP output to the _data plane_ CLF.
```

>    - The HLO should copy audit log policies from  _control plane_ CLF pipelines to the _data plane_,
>      to forward only those logs that will also be forwarded by the data plane CLF.

Reviewer 3: ```suggestion
   - The HLO should copy audit log policies from  _data plane_ CLF pipelines to the _control plane_,
     to forward only those logs that will also be forwarded by the data plane CLF.
```

Customer can set policy in the data plane CLF, it needs copied to the control plane to be applied.

Author: Oops again. Note that the policy will be applied on the data plane, it is an optimization (but a very important one) to apply it on the control plane first to reduce the traffic volume. It is also possible for the control plane to apply a default policy of its own to put an upper bound on what the data plane can request. E.g. the data plan could filter out silly stuff like "get -A pods with full response body" so that the data plane will never get that data even if their filter allows it. Need SRE experties to say what the right defaults are, but the audit policy code has defaults inherited from the old splunk forwarder that might be reasonable to impose.

Reviewer 3: @alanconway where are control plane defaults documented?  How are merges between control plane and data plane policy managed?  What messaging is given to customer if their policy is asking for more than what the control plane will allow?  The UX here is important to understand.  And before building this I'd want to identify the use case.  If customers want all audit events, what is the implications to RH if we allow it in the managed context?  Not sure we even have egress costs, but that may be it.  Load on the MC?

>      to forward only those logs that will also be forwarded by the data plane CLF.

Reviewer 3: Will there be a default set of policies applied if no policy is configured on the data plane?

Author: The control plane can impose such a policy itself, in addition to the data plane policy. You can apply multiple audit policies in the CLF filter chain.
The CLF audit policy has some documented default behavior (inherited from the old splunk audit exporter) but it is possible for the user to over-ride that and ask for 'literally everything' - I think the controle plane probably should impose it's own baseline policy to prevent indiscriminate "give me request bodies for everything" even if the user filter allows it.

Reviewer 3: Control plane forcing a policy is something that won't fit for all use cases, as long as it can be managed per output then it will be OK.

See "API Extensions" and "Implementation Details" for more.

### API Extensions

> None. The HLO watches existing resources: HCP on the control-plane, CLF on the data plane.

Reviewer 2: If there is no new custom resource defined, is it necessary to be an operator here?

Reviewer: How will a workload be deployed in either cluster without something reacting to intent by the admin? Operators are designed to manage a workload based upon a spec. The alternative is the admin crafts every needed resource and deploys it and reverts or modifies it when changes happen to them, intentional or not

Author: @jcantrill this is an unusual case. The HLO operator does not have any resources of its own. It  observes CLF resources  _in another cluster_ which is possible because the data-plane API server actually runs on the management cluster.  It modifies CLF resources in its own cluster to create the HTTP connection from management-CLF to guest-CLF.  It does not actually implement any of the forwarding behavior itself, it just observes and congifgures CLF resources and lets the CLO do it's normal job of enacting the configuration.

### Version Skew Strategy

Data and control plane may have different versions of CLO installed. Points of contact:

- HTTP input-to-output used to forward audit logs.
- API audit policy configuration.

> As long as the HTTP protocol used for log forwarding is stable, the CLO versions can vary.

Author: CLO versions can co-exist provided the HTTP forwarding is compatible at the protocol level (CLO API can change as long as the wire-format does not.)  Protocol changes will ideally be backwards-compat, but need to keep an eye on it if we extend HTTP support (e.g. to support forwarding infra logs)

These are protected by normal backwards-compatibility rules between releases.

### Operational Aspects of API Extensions

The HLO operator watches resources on both control and data planes.
This is unusual but not unprecedented, SREs need to be aware to understand and fix problems.

- 2 new operators to run on management clusters: CLO and HLO.
- Custom Splunk forwarder will be removed.

**Note**: The custom Splunk forwarder MUST be removed and replaced with the CLO for all this to work.

> For non-hypbershift clusters the Splunk exporter and the CLO can coexist by both scraping the API audit log files.

Reviewer 3: ```suggestion
For non-hypershift clusters the Splunk exporter and the CLO can coexist by both scraping the API audit log files.
```

### Version Skew Strategy

Data and control plane may have different versions of CLO installed. Points of contact:

- HTTP input-to-output used to forward audit logs.
- API audit policy configuration.

As long as configuration for these two items remains compatible, the CLO versions can vary.
These are protected by normal backwards-compatibility rules between releases.

It is important to verify and test this compatibility with CLO and HCP releases,

> so we can advertise the correct supported version ranges.

Reviewer 3: Where does QE responsibility for this live?  And specifically in the context of Red Hat Managed HCP?

### Operational Aspects of API Extensions

The HLO operator watches resources on both control and data planes.
This is unusual but not unprecedented, SREs need to be aware to understand and fix problems.

- 2 new operators to run on management clusters: CLO and HLO.
- Custom Splunk forwarder will be removed.

> **Note**: The custom Splunk forwarder MUST be removed and replaced with the CLO for all this to work.

Reviewer 3: Nit, the forwarder can remain but the webhook configuration on the HostedCluster has to be updated to feed to CLF.

Author: Good nit.

> For non-hypbershift clusters the Splunk exporter and the CLO can coexist by both scraping the API audit log files.

Reviewer 3: ```suggestion
For non-hypershift clusters the Splunk exporter and the CLO can coexist by both scraping the API audit log files.
```
